session.go

```diff
@@ -699,9 +703,23 @@ func (q *Query) defaultsFromSession() {
 	q.serialCons = s.cfg.SerialConsistency
 	q.defaultTimestamp = s.cfg.DefaultTimestamp
 	q.idempotent = s.cfg.DefaultIdempotence
+	q.metrics = make(map[string]*queryMetrics)
 	s.mu.RUnlock()
 }
 
+func (q *Query) getHostMetrics(host *HostInfo) *queryMetrics {
+	q.session.mu.Lock()
+	defer q.session.mu.Unlock()
```

Comment:
As there's no way you're returning from this function early, I'd drop the defer and move the mutex unlock down. It's pretty clear so there's no need for the defer overhead, especially because this could be called a sizable number of times while a query is executing and the retry policy consulted. Come to think of it, making this available to the retry policy isn't implemented yet, right?

Reply:
no, not implemented, and I'm not sure retryPolicy needs to deal with metrics at all. But considering the Attempt() gets the whole query, it may, in fact, use the metrics if relevant for the retry policy.